src/System.Net.Mime/src/System/Net/Mime/ContentType.cs

```diff
@@ -0,0 +1,319 @@
+
+        internal void Set(string contentType, HeaderCollection headers)
+        {
+            _type = contentType;
+            ParseValue();
+            headers.InternalSet(MailHeaderInfo.GetString(MailHeaderID.ContentType), ToString());
+            _isPersisted = true;
+        }
+
+        internal void PersistIfNeeded(HeaderCollection headers, bool forcePersist)
+        {
+            if (IsChanged || !_isPersisted || forcePersist)
+            {
+                headers.InternalSet(MailHeaderInfo.GetString(MailHeaderID.ContentType), ToString());
+                _isPersisted = true;
+            }
+        }
+
+        internal bool IsChanged => _isChanged || _parameters != null && _parameters.IsChanged;
+
+        public override string ToString()
+        {
+            if (_type == null || IsChanged)
+            {
+                _type = Encode(false); // Legacy wire-safe format
+                _isChanged = false;
+                _parameters.IsChanged = false;
+                _isPersisted = false;
+            }
+            return _type;
+        }
+
+        internal string Encode(bool allowUnicode)
+        {
+            var builder = new StringBuilder();
+
+            builder.Append(_mediaType); // Must not have unicode, already validated
+            builder.Append('/');
+            builder.Append(_subType);  // Must not have unicode, already validated
+
+            // Validate and encode unicode where required
+            foreach (string key in Parameters.Keys)
+            {
+                builder.Append("; ");
+                EncodeToBuffer(key, builder, allowUnicode);
+                builder.Append('=');
+                EncodeToBuffer(_parameters[key], builder, allowUnicode);
+            }
+
+            return builder.ToString();
+        }
+
+        private static void EncodeToBuffer(string value, StringBuilder builder, bool allowUnicode)
+        {
+            Encoding encoding = MimeBasePart.DecodeEncoding(value);
+            if (encoding != null) // Manually encoded elsewhere, pass through
+            {
+                builder.Append('\"').Append(value).Append('"');
+            }
+            else if ((allowUnicode && !MailBnfHelper.HasCROrLF(value)) // Unicode without CL or LF's
+                || MimeBasePart.IsAscii(value, false)) // Ascii
+            {
+                MailBnfHelper.GetTokenOrQuotedString(value, builder, allowUnicode);
+            }
+            else
+            {
+                // MIME Encoding required
+                encoding = Encoding.GetEncoding(MimeBasePart.DefaultCharSet);
+                builder.Append('"').Append(MimeBasePart.EncodeHeaderValue(value, encoding, MimeBasePart.ShouldUseBase64Encoding(encoding))).Append('"');
+            }
+        }
+
+        public override bool Equals(object rparam) =>
+            rparam == null ? false : string.Compare(ToString(), rparam.ToString(), StringComparison.OrdinalIgnoreCase) == 0;
```

Comment:
Will change.  I made a few tweaks to the code as I was bringing it over, but for the most part I didn't change stuff like this.